### PR TITLE
Port code from NBitcoin and create nuget package

### PR DESCRIPTION
**Pull request by webwarrior-ws**
Ported Litecoin class and all code needed for it from NBitcoin repository.
Added a test project and ported tests from NBitcoin that use Litecoin class.
Added CI workflow that compiles solution and runs tests, generates nuget package and uploads it to nuget.org.

### CODE REVIEW

File: src/NLitecoin/Litecoin.fs [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+            .AddSeeds(NetworkSetBase.ToSeed pnSeed6_main)
+            .SetGenesis("010000000000000000000000000000000000000000000000000000000000000000000000d9ced4ed1130f7b7faad9be25323ffafa33232a17c3edf6cfd97bee6bafbdd97b9aa8e4ef0ff0f1ecd513f7c0101000000010000000000000000000000000000000000000000000000000000000000000000ffffffff4804ffff001d0104404e592054696d65732030352f4f63742f32303131205374657665204a6f62732c204170706c65e280997320566973696f6e6172792c2044696573206174203536ffffffff0100f2052a010000004341040184710fa689ad5023690c80f3a49c8f13f8d45b8c857fbcbc8bc4a8e4d3eb4b10f4d4604fa08dce601aaf0f470216fe1b51850b4acf21b179c45070ac7b03a9ac00000000")
+    
+    /// not used in GWallet
+    override self.CreateTestnet() =
+        null
+
+    /// not used in GWallet
+    override self.CreateRegtest() =
+        null
```
Comment (knocte) [inline src/NLitecoin/Litecoin.fs:350]: @webwarrior-ws better FAIL FAST, which means throw, not null (do this in other parts of the code please)

File: src/NLitecoin/Litecoin.fs [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+            .AddAlias("litecoin-main")
+            .SetUriScheme("litecoin")
+            .AddDNSSeeds(
+                [|
+                    DNSSeedData("loshan.co.uk", "seed-a.litecoin.loshan.co.uk")
+                    DNSSeedData("thrasher.io", "dnsseed.thrasher.io")
+                    DNSSeedData("litecointools.com", "dnsseed.litecointools.com")
+                    DNSSeedData("litecoinpool.org", "dnsseed.litecoinpool.org")
+                    DNSSeedData("koin-project.com", "dnsseed.koin-project.com")
+                |])
```
Comment (knocte) [inline src/NLitecoin/Litecoin.fs]: @webwarrior-ws I don't think geewallet needs DNSSeeds?

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        run: dotnet restore
+      
+      - name: Build
+        run: dotnet build --configuration Release --no-restore
+      
+      - name: Package
+        # https://stackoverflow.com/questions/70249519/how-to-check-if-a-secret-variable-is-empty-in-if-conditional-github-actions
+        env:
+          NUGET_API_KEY: ${{ secrets.NUGET_API_KEY }}
+        if: "${{ env.NUGET_API_KEY != '' }}"
```
Comment (knocte) [inline .github/workflows/CI.yml]: @webwarrior-ws this should only be run if branch is master

Comment (knocte) [reply .github/workflows/CI.yml]: @webwarrior-ws also please add me as co-owner of NLitecoin package in nuget.org please (same username as here)

File: src/NLitecoin/Litecoin.fs (hunk 1 above)
Comment (webwarrior-ws) [reply src/NLitecoin/Litecoin.fs:350]: We can't throw an exception here. `NetworkSetBase`, `Litecoin`'s superclass, will try to initialize all three instances, even if only one is used. Our only option is to return null.

Comment (knocte) [reply src/NLitecoin/Litecoin.fs:350]: didn't we agree that it doesn't need to inherit from NetworkBase?

**Comment by knocte** [comment]
> Removed code that is not used in GWallet.

Why is this in the PR description? In current master branch there's no code at all, so "removed code that is not used in GWallet" is extremely confusing. Please be mindful when writing PR descriptions.

File: src/NLitecoin/Litecoin.fs (hunk 1 above)
Comment (webwarrior-ws) [reply src/NLitecoin/Litecoin.fs:350]: No, it needs to inherit from `NetworkBase`.
See https://github.com/MetacoSA/NBitcoin/blob/93ef4532b9f2ea52b2c910266eeb6684f3bd25de/NBitcoin/NetworkSet.cs#L69C12-L69C12

**Comment by knocte** [comment]
> Why is this in the PR description? In current master branch there's no code at all, so "removed code that is not used in GWallet" is extremely confusing. Please be mindful when writing PR descriptions.

If I explain to you why you have to remove the line "removed code that is not used in GWallet" it is so that you can take the initiative yourself to double-check again all the lines in the description to see if they need to remove as well, otherwise I would just tell you to remove without giving you any reasoning @webwarrior-ws . Please Taras be more proactive.

File: src/NLitecoin/Litecoin.fs (hunk 1 above)
Comment (knocte) [reply src/NLitecoin/Litecoin.fs:350]: @webwarrior-ws but is NetworkBase used in geewallet? I believe we already had this conversation and you said that it is not

Comment (webwarrior-ws) [reply src/NLitecoin/Litecoin.fs:350]: Not directly. But it is used by `NetworkBuilder` to build `Network`, and that **is** used in GWallet.

Comment (knocte) [reply src/NLitecoin/Litecoin.fs:350]: ok fair enough; I'm then marking this convo as resolved with the `Resolve conversation` button

File: .github/workflows/CI.yml (hunk 3 above)
Comment (webwarrior-ws) [reply .github/workflows/CI.yml]: Done

File: src/NLitecoin/Litecoin.fs (hunk 2 above)
Comment (webwarrior-ws) [reply src/NLitecoin/Litecoin.fs]: Indeed GWallet seems to do fine without DNSSeeds. Removed them.

**Comment by knocte** [comment]
Last change to be done here before I merge: pull first (to get my commit) with --rebase, and after that squash the commits that change FSharp.Core version into 1, because there's no point to split them (given that the oldest one didn't achieve what you wanted).

**Comment by webwarrior-ws** [comment]
Squashed version commits, but didn't use push1by1 to not trigger creation of several package versions.

**Comment by knocte** [comment]
> Squashed version commits, but didn't use push1by1 to not trigger creation of several package versions.

The way to avoid that problem in the future is to never propose a PR from a branch called "master"; the branch name should always reflect what you're working on (and this way pushing to it doesn't push to nuget).

File: tests/NLitecoin.Tests/Usings.cs [hunk 4]
```diff
@@ -0,0 +1 @@
+global using NUnit.Framework;
```
Comment (knocte) [inline tests/NLitecoin.Tests/Usings.cs]: @webwarrior-ws let's remove this file, I'm not a fan

**Comment by knocte** [comment]
@webwarrior-ws actually how did you check geewallet running against this, just checking balances? don't tell me you haven't tested a transaction and still think this should be working?

File: tests/NLitecoin.Tests/Usings.cs (hunk 4 above)
Comment (webwarrior-ws) [reply tests/NLitecoin.Tests/Usings.cs]: Removed Usings.cs

**Comment by webwarrior-ws** [comment]
> @webwarrior-ws actually how did you check geewallet running against this, just checking balances? don't tell me you haven't tested a transaction and still think this should be working?

I ran unit tets, integration test, ran the console app. I didn't perform any transactions.

**Comment by knocte** [comment]
>  I didn't perform any transactions.

Well, wallets are for performing transactions. I don't know how would you think only testing read-only operations is enough. If transactions don't work we're essentially freezing the funds of our users.

**Comment by webwarrior-ws** [comment]
Just tested sending some LTC. It worked.

**Comment by knocte** [comment]
Ok, now can you please push 1by1, but to avoid it publishing many nuget pkgs, just move the commit that adds nugetPush to be the last one. (And now that you're going to change order of commits, you could also make the one that adds CI the first one.)

**Comment by knocte** [comment]
> now can you please push 1by1,

After that, you can start looking into Milestone 2. For that, I recommend using Litecoin-Core (with regtest or testnet) to create test transactions, that should later try to parse with NLitecoin. The goal is to have geewallet be able to send MimbleWimble-style transactions in Litecoin (we will investigate later about the UX of how to expose the feature to the user, but for now just make it work).

**Comment by knocte** [comment]
BTW the FSharpCore commit contains a uint->uint32 change, is that intentional?

**Comment by webwarrior-ws** [comment]
> BTW the FSharpCore commit contains a uint->uint32 change, is that intentional?

Yes, older F# doesn't have `uint` alias, only `uint32`. So I had to change it.

**Comment by knocte** [comment]
> Yes, older F# doesn't have uint alias, only uint32

Very interesting. Next time please clarify it in commit message because it's not immediately obvious.

BTW one of the commits has red CI, you might want to squash it with next?

**Comment by webwarrior-ws** [comment]
> BTW one of the commits has red CI, you might want to squash it with next?

You told me to "make the one that adds CI the first one", and I did so. That commit also introduces tests, so it fails, because there is no code to test yet.

**Comment by knocte** [comment]
Fair enough but following what happened at the conventions PR, it should contain the term "failing test" ;-)

Merging.

**Pull request merged**